src/backend/distributed/planner/insert_select_planner.c

```diff
@@ -374,12 +374,36 @@ CreateInsertSelectIntoLocalTablePlan(uint64 planId, Query *originalQuery, ParamL
 									 boundParams, bool hasUnresolvedParams,
 									 PlannerRestrictionContext *plannerRestrictionContext)
 {
-	RangeTblEntry *selectRte = ExtractSelectRangeTableEntry(originalQuery);
+	Query *insertSelectQuery = copyObject(originalQuery);
 
-	Query *selectQuery = BuildSelectForInsertSelect(originalQuery);
-	originalQuery->cteList = NIL;
-	DistributedPlan *distPlan = CreateDistributedPlan(planId, selectQuery,
-													  copyObject(selectQuery),
+	RangeTblEntry *selectRte = ExtractSelectRangeTableEntry(insertSelectQuery);
+	RangeTblEntry *insertRte = ExtractResultRelationRTEOrError(insertSelectQuery);
+	Oid targetRelationId = insertRte->relid;
+
+	ReorderInsertSelectTargetLists(insertSelectQuery, insertRte, selectRte);
+
+	/*
+	 * Cast types of insert target list and select projection list to
+	 * match the column types of the target relation.
+	 */
+	selectRte->subquery->targetList =
+		AddInsertSelectCasts(insertSelectQuery->targetList,
+							 selectRte->subquery->targetList,
+							 targetRelationId);
+
+	/*
+	 * Due to ReorderInsertSelectTargetLists(), now the Vars in GROUP BY
+	 * clause might be pointing to an incorrect range table entry and they
+	 * matter for the local execution. For this reason, need to wrap SELECT
+	 * query in a subquery in that case, in addition to the cases where
+	 * BuildSelectForInsertSelect() does so by default.
+	 */
+	bool wrapIfContainsGroupBy = true;
+	selectRte->subquery = BuildSelectForInsertSelect(insertSelectQuery,
```

Comment:
BuildSelectForInsertSelect has a comment that it wraps top-level set operations as to not confuse ReorderInsertSelectTargetLists.

Given that we now do ReorderInsertSelectTargetLists first, what happens to top-level set operations? (insert into ... select .. union select ..)

Reply:
Seems that calling BuildSelectForInsertSelect() later is also fine.

```sql
[local] onurctirtir@onurctirtir:10700-4363=# CREATE TABLE local_table_5(
      col_1 integer,
      col_2 integer,
      col_3 text,
      col_4 text default 'deftext',
      col_5 int,
      col_6 text,
      col_7 text,
      col_8 text
  );
table_1(
    dist_col integer,
    int_col integer,
    text_col_1 text,
    text_col_2 text
);
SELECT create_distributed_table('dist_table_1', 'dist_col');
INSERT INTO dist_table_1 VALUES (10, 11, 'sCREATE TABLE
Time: 26.014 ms
[local] onurctirtir@onurctirtir:10700-4363=# 
[local] onurctirtir@onurctirtir:10700-4363=# CREATE TABLE dist_table_1(
      dist_col integer,
      int_col integer,
      text_col_1 text,
      text_col_2 text
  );
CREATE TABLE
Time: 9.212 ms
[local] onurctirtir@onurctirtir:10700-4363=# SELECT create_distributed_table('dist_table_1', 'dist_col');
┌──────────────────────────┐
│ create_distributed_table │
├──────────────────────────┤
│                          │
└──────────────────────────┘
(1 row)

Time: 255.095 ms
[local] onurctirtir@onurctirtir:10700-4363=# INSERT INTO dist_table_1 VALUES (10, 11, 'string_10', 'string_11');
INSERT 0 1
Time: 14.637 ms
[local] onurctirtir@onurctirtir:10700-4363=# INSERT INTO local_table_5 (col_3)                                                                                                           
SELECT t1.text_col_1
FROM dist_table_1 t1
JOIN dist_table_1 t2
ON (t1.dist_col = t2.dist_col)
GROUP BY t1.text_col_1;
INSERT 0 1
Time: 32.959 ms
[local] onurctirtir@onurctirtir:10700-4363=# INSERT INTO local_table_5 (col_3)                                                                                                  
SELECT 'my_text'
FROM dist_table_1 t1
JOIN dist_table_1 t2
ON (t1.dist_col = t2.dist_col)
GROUP BY t1.text_col_1;
INSERT 0 1
Time: 18.112 ms
[local] onurctirtir@onurctirtir:10700-4363=# table local_table_5 ;
┌───────┬───────┬───────────┬─────────┬───────┬───────┬───────┬───────┐
│ col_1 │ col_2 │   col_3   │  col_4  │ col_5 │ col_6 │ col_7 │ col_8 │
├───────┼───────┼───────────┼─────────┼───────┼───────┼───────┼───────┤
│       │       │ string_10 │ deftext │       │       │       │       │
│       │       │ my_text   │ deftext │       │       │       │       │
└───────┴───────┴───────────┴─────────┴───────┴───────┴───────┴───────┘
(2 rows)
```

Reply:
hmm, true that set operations are broken now:

```sql
INSERT INTO local_table_5 (col_5, col_3)
SELECT 12, 'string_11' FROM dist_table_1 t1 GROUP BY t1.text_col_1
UNION
SELECT int_col, 'string' FROM dist_table_1;
```